tests/integration/test_tls.py

```diff
@@ -5,51 +5,161 @@
 from pytest_operator.plugin import OpsTest
 
 from tests.integration.helpers.helpers import (
+    CLIENT_APP_NAME,
     PGB,
     PGB_METADATA,
     POSTGRESQL_APP_NAME,
+    app_name,
+    check_tls,
     deploy_and_relate_application_with_pgbouncer,
+    scale_application,
 )
 
-MATTERMOST_APP_NAME = "mattermost"
+MATTERMOST_APP_NAME = "mattermost-k8s"
 TLS_CERTIFICATES_APP_NAME = "tls-certificates-operator"
 APPLICATION_UNITS = 2
 DATABASE_UNITS = 3
 
 
+@pytest.mark.abort_on_fail
 @pytest.mark.tls_tests
-async def test_mattermost_db(ops_test: OpsTest) -> None:
-    """Deploy Mattermost to test the 'db' relation.
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build and deploy pgbouncer charm."""
+    wait_for_apps = False
 
-    Mattermost needs TLS enabled on PgBouncer to correctly connect to it.
+    if not await app_name(ops_test):
+        wait_for_apps = True
+        charm = await ops_test.build_charm(".")
+        async with ops_test.fast_forward():
+            await ops_test.model.deploy(
+                charm,
+                resources={
+                    "pgbouncer-image": PGB_METADATA["resources"]["pgbouncer-image"][
+                        "upstream-source"
+                    ]
+                },
+                application_name=PGB,
+                num_units=APPLICATION_UNITS,
+            )
 
-    Args:
-        ops_test: The ops test framework
-    """
-    charm = await ops_test.build_charm(".")
-    async with ops_test.fast_forward():
-        await ops_test.model.deploy(
-            charm,
-            resources={
-                "pgbouncer-image": PGB_METADATA["resources"]["pgbouncer-image"]["upstream-source"]
-            },
-            application_name=PGB,
-            num_units=APPLICATION_UNITS,
+    if not await app_name(ops_test, CLIENT_APP_NAME):
+        wait_for_apps = True
+        application_charm = await ops_test.build_charm(
+            "tests/integration/relations/pgbouncer_provider/application-charm"
+        )
+        async with ops_test.fast_forward():
+            await ops_test.model.deploy(
+                application_charm,
+                application_name=CLIENT_APP_NAME,
+            )
+    # remove preexisting relation if any so that we can know the rel id
+    relations = [
+        relation
+        for relation in ops_test.model.applications[PGB].relations
+        if not relation.is_peer
+        and f"{relation.requires.application_name}:{relation.requires.name}"
+        == f"{CLIENT_APP_NAME}:first-database"
+    ]
+    if relations:
+        await ops_test.model.applications[PGB].remove_relation(
+            f"{PGB}:database", f"{CLIENT_APP_NAME}:first-database"
         )
+        await ops_test.model.wait_for_idle(status="active", timeout=1000)
```

Comment:
This is not strictly necessary, but it helps with running the tests consistently against an existing cluster